numba/runtime/nrt.c

```diff
@@ -52,8 +64,77 @@ typedef struct {
 
 
 static
-MemInfo *nrt_pop_meminfo_list(MemInfo * volatile *list) {
+void list_lock(MemInfoList *listst, word_type mask) {
+    volatile word_type actual;
+    word_type old;
+    word_type newval;
+
+    /* First load */
+    actual = listst->state;
+
+    while (1) {
+        /* `actual` is loaded in the loop by atomic_cas */
+        old = actual;
+
+        /* If in unlocked state, set to init "mask" state */
+        if ( old == UNLOCK ) newval = mask;
+        /* If in mask state, increment */
+        else if ( (old & mask) == mask
+                  && (old & LOCK_COUNT_MASK) < LOCK_COUNT_MASK )
+            newval = old + 1;
+        /* If not in mask state, load and retry */
+        else {
+            actual = listst->state;
+            continue;
+        }
+
+        /* CAS with atomic load to actual */
+        if (TheMSys.atomic_cas(&listst->state, old, newval, &actual)) {
+            /* Succeed, quit */
+            return;
+        }
+    }
+}
+
+static
+void list_unlock(MemInfoList *listst, word_type mask) {
+    volatile word_type actual;
+    word_type old;
+    word_type newval;
+
+    /* First load */
+    actual = listst->state;
+
+    while (1) {
+        /* `actual` is loaded in the loop by atomic_cas */
+        old = actual;
+
+        /* If in init "mask" state, set to unlock state */
+        if ( old == mask ) newval = UNLOCK;
+        /* If in mask state, decrement */
+        else if ( (old & mask) == mask
+                   && (old & LOCK_COUNT_MASK) <= LOCK_COUNT_MASK)
+           newval = old - 1;
+        /* If non "mask" state, fatal error */
+        else {
+            puts("Corrupted list state");
+            exit(1);
+        }
+
+        /* CAS with atomic load to `actual` */
+        if (TheMSys.atomic_cas(&listst->state, old, newval, &actual)) {
+            /* Succeed, quit */
+            return;
+        }
+    }
+}
+
+static
+MemInfo *nrt_pop_meminfo_list(MemInfoList *listst) {
     MemInfo *old, *repl, *head;
+    MemInfo * volatile * volatile list = &listst->list;
+
+    list_lock(listst, LOCK_POP);
```

Comment:
This looks complicated and difficult to understand. Why can't you lock the list in an exclusive way and then do stuff on it unconditionally? It looks like you are both locking the list *and* still allowing concurrent modifications to it. This seems hard to reason about.